enhancements/network/ovn-hypershift-live-migration.md

```diff
@@ -0,0 +1,314 @@
+└────────────────────┘            └─────────────────────┘
+```
+
+**VMs logical switch ports:**
+
+The logical switch port for new VMs will use ovn-k ipam to reserve an IP address,
+and when live-migrating the VM, the LSP address will be re-used.
+
+CNI must avoid setting the IP address on the migration destination pod but ovn-k controllers should 
+continue book keeping it.
+
+Also the DHCP options will be configured to deliver the address to the VMs
+
+```text
+    ┌──────────────────────┐   ┌────────────────────┐
+    │logical switch node1  │   │logical switch node2│
+    └──────────────────────┘   └────────────────────┘
+┌──────────────────────┐  │     │   ┌──────────────────────┐      
+│ lsp ns-virt-launcher1│──┘     └───│ lsp ns-virt-launcher2│     
+│ dhcpv4_options: 1234 │            │ dhcpv4_options: 1234 │
+│ address:             │            │ address:             │    
+│  0a:58:0a:f4:00:01   │            │  0a:58:0a:f4:00:01   │
+│  10.244.0.8          │            │  10.244.0.8          │
+└──────────────────────┘            └──────────────────────┘
+
+┌─────────────────────────────────┐
+│ dhcp-options 1234               │
+│   lease_time: 3500              │
+│   router: 169.254.1.1           │
+│   dns_server: [kubedns]         │
+│   server_id: 169.254.1.1        │
+│   server_mac: c0:ff:ee:00:00:01 │
+└─────────────────────────────────┘
+```
+
+**Virt-launcher pod address:**
+The CNI will not set an address at virt-launcher pod netns, that address is
+assigned to the VM with the DHCP options from the LSP, this allows to use 
+kubevirt bridge binding with pod networking and still do live migration.
+
+#### IPAM
+
+The point to point routing feature allows an address to be running at a node different from
+the one "owning" the subnet the address is coming from. This will happen
+after VM live migration.
+
+One scenario for live migration is to shut down the node the VMs were migrated
+from, this means that the IPAM node subnet should go back to the pool but since
+the migrated VMs contains IPs from it those IPs should reserved in case the 
+subnet is assigned to a new node.
+
+On that case before assigning the subnet to the node the VMs ips need to be 
+reserved so they don't get assigned to new pods
+
+Another scenario is ovn-kubernetes pods restarting after live migration, on 
+that case ovnkube-master should discover to what IP pool the VM belongs 
+(using an annotation or searching for subnets) and reserve the address.
+
+#### Detecting hypershift VMs and live migration
+
+This enhancement is implemented as part of ovn-k pod's CNI and controllers so
+the feature detection mechanism have to be implemented as part of kubevirt
+virt-launcher pods.
+
+To skip CNI IP configuration and assign assign DHCP options to LSP the following
+label and annotation are check to ensure that only kubevirt pods with explicit configuration
+are affected:
+- `ovn.org/skip-ip-configuration-on-cni`
+- `kubevirt.io/vm`
```

Comment:
I think there's no real need for checking this annotation - other than limiting the feature to kubevirt VMs.

This could be used to implement "sticky" IPs on stateful sets for instance. 

But I am not opposing; we could start by limiting the feature to kubevirt VMs, and make it more general in the future.

Reply:
That's the idea, let's limiting it at the beggining and remove de limitation at follow up with proper testing.